components/gitpod-protocol/go/gitpod-service.go

```diff
@@ -336,21 +338,51 @@ func (gp *APIoverJSONRPC) InstanceUpdates(ctx context.Context, instanceID string
 	chn := make(chan *WorkspaceInstance)
 
 	gp.mu.Lock()
-	if gp.subs == nil {
-		gp.subs = make(map[string]map[chan *WorkspaceInstance]struct{})
+	if gp.instanceSubs == nil {
+		gp.instanceSubs = make(map[string]map[chan *WorkspaceInstance]struct{})
 	}
-	if sub, ok := gp.subs[instanceID]; ok {
+	if sub, ok := gp.instanceSubs[instanceID]; ok {
 		sub[chn] = struct{}{}
 	} else {
-		gp.subs[instanceID] = map[chan *WorkspaceInstance]struct{}{chn: {}}
+		gp.instanceSubs[instanceID] = map[chan *WorkspaceInstance]struct{}{chn: {}}
 	}
 	gp.mu.Unlock()
 
 	go func() {
 		<-ctx.Done()
 
 		gp.mu.Lock()
-		delete(gp.subs[instanceID], chn)
+		delete(gp.instanceSubs[instanceID], chn)
+		close(chn)
+		gp.mu.Unlock()
+	}()
+
+	return chn, nil
+}
+
+// WorkspaceUpdates subscribes to workspace instance updates until the context is canceled
+func (gp *APIoverJSONRPC) WorkspaceUpdates(ctx context.Context, workspaceID string) (<-chan *WorkspaceInstance, error) {
```

Comment:
A possibly nicer return interface would be something like an `Iterator` with `Next() (val, closed)` and `Close()`.

Reply:
This would also help with the issue below where two subscriptions for the same workspace ID end up with side-effects

Reply:
Can we do this later, I'm replacing the old `InstanceUpdates` so making this change in the interface could be risky and will take more time to test so it doesn't break other components